File: app/services/integrations/slack_client.rb
module Integrations
  class SlackClient < BaseClient
    API_URL = "https://slack.com/api"

    def test_connection
      response = api_request("auth.test")

      if response["ok"]
        { success: true, message: "Connected to workspace: #{response['team']}" }
      else
        { success: false, message: response["error"] || "Unknown error" }
      end
    rescue => e
      { success: false, message: e.message }
    end

    def sync
      # Slack primarily uses webhooks/events API
      # This method can be used to fetch recent messages from configured channels
      channels = credentials["channels"] || []
      created_count = 0

      channels.each do |channel_id|
        messages = fetch_channel_history(channel_id)

        messages.each do |message|
          next if message["bot_id"].present?
          next unless should_capture_message?(message)

          external_id = "#{channel_id}-#{message['ts']}"
          next if Feedback.find_by_external_id("slack", external_id)

          text = message["text"] || ""
          Feedback.create!(
            source: :slack,
            source_external_id: external_id,
            title: text.truncate(100),
            content: text,
            author_name: message["user"],
            raw_data: message
          )
          created_count += 1
        end
      end

      integration.mark_synced!
      { success: true, created: created_count }
    rescue => e
      { success: false, message: e.message }
    end

    private

    def api_request(method, params = {})
      response = Faraday.post("#{API_URL}/#{method}") do |req|
        req.headers["Authorization"] = "Bearer #{credentials['bot_token']}"
        req.headers["Content-Type"] = "application/json"
        req.body = params.to_json
      end

      JSON.parse(response.body)
    end

    def fetch_channel_history(channel_id)
      oldest = integration.last_synced_at&.to_i || 24.hours.ago.to_i

      response = api_request("conversations.history", {
        channel: channel_id,
        oldest: oldest.to_s,
        limit: 100
      })

      response["ok"] ? response["messages"] : []
    end

    def should_capture_message?(message)
      text = message["text"]&.downcase || ""
      keywords = credentials["keywords"] || %w[feedback bug issue problem feature request]
      keywords.any? { |keyword| text.include?(keyword.downcase) }
    end
  end
end

File: apps/worker/src/integrations/slack-client.ts
import { BaseIntegrationClient, type FeedbackItem, type TestConnectionResult } from "./base-client.js";

export class SlackClient extends BaseIntegrationClient {
  async testConnection(): Promise<TestConnectionResult> {
    const token = this.str("bot_token");
    if (!token) return { success: false, message: "Bot token is required" };

    try {
      const res = await fetch("https://slack.com/api/auth.test", {
        headers: { Authorization: `Bearer ${token}` },
      });
      const json = (await res.json()) as { ok: boolean; error?: string; team?: string };
      if (!json.ok) return { success: false, message: json.error ?? "Auth test failed" };
      return { success: true, message: `Connected to ${json.team ?? "workspace"}` };
    } catch (err) {
      return { success: false, message: err instanceof Error ? err.message : "Connection failed" };
    }
  }

  async fetchItems(): Promise<FeedbackItem[]> {
    const token = this.str("bot_token");
    const channel = this.str("channel_id") || this.str("channels");
    console.log(`[slack-sync] creds keys=${Object.keys(this.credentials).join(",")} token=${token ? "set" : "MISSING"} channel=${channel || "MISSING"} raw_channels=${this.str("channels")}`);
    if (!token || !channel) return [];

    const oldest = String(Math.floor((Date.now() - 60 * 60 * 1000) / 1000)); // last hour

    const res = await fetch(`https://slack.com/api/conversations.history?channel=${channel}&oldest=${oldest}&limit=100`, {
      headers: { Authorization: `Bearer ${token}` },
    });

    const json = (await res.json()) as { ok: boolean; messages?: SlackMessage[]; error?: string };
    console.log(`[slack-sync] channel=${channel} ok=${json.ok} error=${json.error ?? "none"} messages=${json.messages?.length ?? 0}`);
    if (!json.ok || !json.messages) return [];

    const items: FeedbackItem[] = [];
    for (const msg of json.messages) {
      if (!msg.text) continue;
      // Skip Slack system events (joins, topic changes, pinned items, etc.) —
      // standard user messages have no `subtype`.
      if (msg.subtype) continue;

      items.push({
        title: msg.text.slice(0, 200),
        content: msg.text,
        authorName: msg.user ?? undefined,
        sourceExternalId: `${channel}-${msg.ts}`,
        rawData: msg as unknown as Record<string, unknown>,
      });
    }

    return items;
  }
}

interface SlackMessage {
  type?: string;
  subtype?: string;
  user?: string;
  text?: string;
  ts?: string;
}
